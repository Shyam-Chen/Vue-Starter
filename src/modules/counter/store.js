import { provide, inject, reactive, computed } from 'vue';

const stateSymbol = Symbol('counter');

const createState = reactive({
  count: 0,
});

export const useState = () => inject(stateSymbol);

export const useActions = () => {
  const state = useState();

  return {
    increment() {
      state.count += 1;
    },
    decrement() {
      state.count -= 1;
    },
  };
};

export const useComputeds = () => {
  const state = useState();

  return {
    evenOrOdd: computed(() => (state.count % 2 === 0 ? 'Even' : 'Odd')),
  };
};

export default () => provide(stateSymbol, createState);
