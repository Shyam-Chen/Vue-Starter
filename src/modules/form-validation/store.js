import { provide, inject, reactive, computed } from 'vue';

const stateSymbol = Symbol('counter');

const createState = reactive({
  signInForm: {
    username: '',
    password: '',
  },
  errors: {},
});

export const useState = () => inject(stateSymbol);

export const useActions = () => {
  const state = useState();

  return {
    signIn() {
      console.log('Sign-in');
    },
  };
};

export const useComputeds = () => {
  const state = useState();

  return {};
};

export default () => provide(stateSymbol, createState);
