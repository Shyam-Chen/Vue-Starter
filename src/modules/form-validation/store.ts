import { reactive, inject } from 'vue';

import { defineContext } from '~/composables';

import type { State } from './types';

export const stateSymbol = Symbol('/form-validation');

export const createState = reactive({
  basicForms: {},

  errors: {},
});

export const useState = () => inject(stateSymbol) as State;

export const useActions = () => {
  const state = useState();

  const actions = {
    signIn() {
      console.log('Sign In', state.basicForms);
    },
  };

  return actions;
};

export const useComputeds = () => {
  const state = useState();

  const computeds = {};

  return computeds;
};

export default () => defineContext(stateSymbol, createState);
