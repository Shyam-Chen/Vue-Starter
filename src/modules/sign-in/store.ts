import { reactive, inject } from 'vue';
import { useRouter } from 'vue-router';

import { defineContext, useSysFetch } from '~/composables';

import type { State } from './types';

export const stateSymbol = Symbol('/sign-in');

export const createState = reactive({
  signInForm: {},
  signedIn: false,

  errors: {},
});

export const useState = () => inject(stateSymbol) as State;

export const useActions = () => {
  const router = useRouter();
  const signInApi = useSysFetch('/sign-in', { immediate: false }).json();
  const state = useState();

  const actions = {
    async signIn() {
      state.signedIn = true;

      await signInApi.post(state.signInForm).execute();

      if (signInApi.statusCode.value === 200) {
        const { token } = signInApi.data.value;
        localStorage.setItem('token', token);
        await router.push('/dashboard');
        state.signedIn = false;
      }
    },
  };

  return actions;
};

export const useComputeds = () => {
  const state = useState();

  const computeds = {};

  return computeds;
};

export default () => defineContext(stateSymbol, createState);
