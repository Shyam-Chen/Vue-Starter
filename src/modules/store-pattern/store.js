import { computed, reactive, watch, provide, inject } from 'vue';

const stateSymbol = Symbol('counter');

const createState = reactive({
  count: 0,
  prevCount: null,
});

export const useState = () => inject(stateSymbol);

export const useActions = () => {
  const state = useState();

  const actions = {
    increment() {
      state.count += 1;
    },
    decrement() {
      state.count -= 1;
    },
    setCount(val) {
      state.count = val;
    },
    resetCount() {
      actions.setCount(0);
    },
  };

  return actions;
};

export const useComputeds = () => {
  const state = useState();

  const computeds = {
    evenOrOdd: computed(() => (state.count % 2 === 0 ? 'Even' : 'Odd')),
  };

  return computeds;
};

export const useReactions = () => {
  const state = useState();

  watch(
    () => state.count,
    (val, oldVal) => {
      state.prevCount = oldVal;
    },
  );
};

export default () => provide(stateSymbol, createState);
