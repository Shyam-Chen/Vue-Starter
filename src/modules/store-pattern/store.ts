import { computed, reactive, watch, provide, inject } from 'vue';

import type { State } from './type';

export const stateSymbol = Symbol('counter');

export const createState = reactive({
  count: 0,
  prevCount: null,
});

export const useState = () => inject(stateSymbol) as State;

export const useActions = () => {
  const state = useState();

  const actions = {
    increment() {
      state.count += 1;
    },
    decrement() {
      state.count -= 1;
    },
    setCount(val: number) {
      state.count = val;
    },
    resetCount() {
      actions.setCount(0);
    },
  };

  return actions;
};

export const useComputeds = () => {
  const state = useState();

  const computeds = {
    evenOrOdd: computed(() => (state.count % 2 === 0 ? 'Even' : 'Odd')),
  };

  return computeds;
};

export const useReactions = () => {
  const state = useState();

  watch(
    () => state.count,
    (val, oldVal) => {
      state.prevCount = oldVal;
    },
  );
};

export default () => provide(stateSymbol, createState);
