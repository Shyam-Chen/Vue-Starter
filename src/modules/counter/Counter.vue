<script setup>
import Display from './Display.vue';
import Buttons from './Buttons.vue';
import store from './store.js';

store();
</script>

<template>
  <Display />
  <Buttons />
</template>
