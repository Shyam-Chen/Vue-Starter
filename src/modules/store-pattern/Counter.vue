<script lang="ts" setup>
import { useState, useActions, useComputeds, useReactions } from './store';

const state = useState();
const actions = useActions();
const computeds = useComputeds();
useReactions();
</script>

<template>
  <div>Clicked: {{ state.count }} times, value is {{ computeds.evenOrOdd.value }}.</div>
  <div>{{ state.prevCount }}</div>

  <button
    class="bg-blue-500 hover:bg-blue-700 text-white px-4 py-1 rounded"
    @click="actions.increment"
  >
    Increment
  </button>
  <button
    class="bg-blue-500 hover:bg-blue-700 text-white px-4 py-1 rounded"
    @click="actions.decrement"
  >
    Decrement
  </button>
</template>
