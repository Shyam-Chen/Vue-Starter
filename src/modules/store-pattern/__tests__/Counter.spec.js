import { test, expect } from 'vitest';
import { render } from '@testing-library/vue';

import { stateSymbol, createState } from '../store';
import Counter from '../Counter.vue';

test('Counter', () => {
  const { html } = render(Counter, { global: { provide: { [stateSymbol]: createState } } });
  expect(html()).toMatchSnapshot();
});
