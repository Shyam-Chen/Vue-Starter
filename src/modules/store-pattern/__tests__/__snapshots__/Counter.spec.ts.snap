// Vitest Snapshot v1

exports[`Counter 1`] = `
"<div>Clicked: 0 times, value is Even.</div>
<div></div>
<button class=\\"bg-blue-500 hover:bg-blue-700 text-white px-4 py-1 rounded\\"> Increment </button>
<button class=\\"bg-blue-500 hover:bg-blue-700 text-white px-4 py-1 rounded\\"> Decrement </button>"
`;
